### ISSUE TITLE
Reduce docker image layers

### ISSUE THREAD

**Issue by sils** [coala/docker-coala-base#59]
might help for https://github.com/coala/docker-coala-base/issues/55

**Comment by sils** [comment]
certainly makes pulling faster

**Comment by yukiisbored** [comment]
It can be done by merging some `RUN` instructions into one but it doesn't help that much since there's a one big image layer. I think it's when the distro is installing the dependencies. Changing into a smaller distro can make a significant change in terms of speed and size especially with that huge chunk but that's quite a huge route.

**Comment by arush0311** [comment]
What about an option of selecting only those bears which the user wants, according to my discussion with @sils we could make a web app which compiles a dockerimage of only those bears which the user wants. Also see these http://stackoverflow.com/questions/24482822/how-to-share-my-docker-image-without-using-the-docker-hub

**Comment by arush0311** [comment]
We could use base image as the coala image with all bears installed and then remove all the bears which are not needed, hence also reducing the time to compile the image

**Comment by yukiisbored** [comment]
but that increases the amount of image layers. Images in Docker are basically layers of changes that compiles into one docker image. And that also requires a server with a good amount of performance and bandwidth since some people just throw it in CI.

**Comment by arush0311** [comment]
@yukiisbored is there a way to remove what was added in previous layers? If that is not the case we will have to start with the base image (alpine?) and then install bears on top of it.

**Comment by arush0311** [comment]
Perhaps we can calculate the shortest (installation) path between the previous images and the current image to be compiled and then use those steps.

**Comment by yukiisbored** [comment]
@arush0311 as far as I know, nope. If it's possible, that'll increase the amount of layers anyway since it requires seperating installation commands

**Comment by arush0311** [comment]
@yukiisbored yes you are right :sweat: Perhaps will help with https://github.com/coala/docker-coala-base/issues/55 though.

**Comment by sils** [comment]
@razvanch did some work in that direction which was merged, is this still needed? Looks ok to me.

**Comment by sils** [comment]
Reopen if you disagree :)

**Comment by jayvdb** [comment]
Still lots of unnecessary layers.

**Comment by jayvdb** [comment]
> What about an option of selecting only those bears which the user wants

@yukiisbored found a nice tool which allows building a custom docker image by selecting a subset of components. @yukiisbored can you create a new issue about that.

**Comment by jayvdb** [comment]
At  https://github.com/coala/docker-coala-base/pull/119#issuecomment-282401596, @fneu  suggested using one script for everything to reduce the layers.

With https://github.com/jayvdb/docker-coala-base/commit/3166a301b, I merged all the commands into a single `RUN` command, and [the result](https://travis-ci.org/jayvdb/docker-coala-base/builds/205461932) is a 1Mb increase over [master](https://travis-ci.org/coala/docker-coala-base/builds/205124492). While I doubt the increase is significant, or even unavoidable, it does indicate that it isnt an automatic win to collapse all of the layers.

**Comment by fneu** [comment]
Unfortunate, but thanks for trying
